src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java

```diff
@@ -0,0 +1,412 @@
+     * handled by calling routine.
+     *
+     * @param vc  Input VC with variants to left align
+     * @param ref Reference context
+     * @return new VC.
+     */
+    protected VariantContext leftAlign(final VariantContext vc, final ReferenceContext ref) {
+        if (!vc.isSimpleIndel()) {
+            return vc;
+        }
+
+        // get the indel length
+        final int indelLength = Math.abs(vc.getIndelLengths().get(0));
+
+        int leadingBases = 50;
+
+        while (leadingBases < maxLeadingBases) {
+            ref.setWindow(leadingBases, maxIndelSize);
+            final byte[] refSeq = ref.getBases();
+
+            // create an indel haplotype.
+            //
+            final int originalIndex = vc.getStart() - ref.getWindow().getStart() + 1;
+            final byte[] originalIndel = makeHaplotype(vc, refSeq, originalIndex, indelLength);
+            // create a CIGAR string to represent the event
+            ArrayList<CigarElement> elements = new ArrayList<CigarElement>();
+            elements.add(new CigarElement(originalIndex, CigarOperator.M));
+            elements.add(new CigarElement(indelLength, vc.isSimpleDeletion() ? CigarOperator.D : CigarOperator.I));
+            elements.add(new CigarElement(refSeq.length - originalIndex, CigarOperator.M));
+            Cigar originalCigar = new Cigar(elements);
+
+            // align no further left than base after previous variant
+            int leftmostAllowedAlignment = furthestEndOfEarlierVariant - ref.getWindow().getStart() + 2;
+            // left align the CIGAR
+            Cigar newCigar = AlignmentUtils.leftAlignIndel(originalCigar, refSeq, originalIndel, 0, 0, leftmostAllowedAlignment, true);
+            if (newCigar.equals(originalCigar)) {
+                return vc;
+            }
+            if (newCigar.getCigarElement(0).getLength() != refSeq.length && newCigar.getCigarElement(0).getOperator() == CigarOperator.M) {
+                int difference = originalIndex - newCigar.getCigarElement(0).getLength();
+                VariantContext newVC = new VariantContextBuilder(vc).start(vc.getStart() - difference).stop(vc.getEnd() - difference).make();
+
+                final int indelIndex = originalIndex - difference;
+                final byte[] newBases = new byte[indelLength + 1];
+                newBases[0] = refSeq[indelIndex - 1];
+                System.arraycopy((vc.isSimpleDeletion() ? refSeq : originalIndel), indelIndex, newBases, 1, indelLength);
+                final Allele newAllele = Allele.create(newBases, vc.isSimpleDeletion());
+                newVC = updateAllele(newVC, newAllele);
+                // overwrite default return value with new left-aligned VC
+                return newVC;
+            }
+            // if we have left aligned to the beginning of the reference window, double leadingBases length and try again
+            leadingBases *= 2;
```

Comment:
It seems to me (first based on the actual tests that hit this line and the above code) that this line will only get hit in the case where the reference is a repetition of a single base, as that seems to be the only likely case where a deletion or insertion will "fall off" a read, and it seems to pick this up. But I see no tests of this auto-expansion behavior working for repetitive units like short tandem repeats, indeed it would seem that the number of units in a repetitive group would need to be divisible by 50 to ever reach this point in the code, otherwise the above left aligning code will just stop and place the indel somewhere short of the last base at the front of the read. Its a shame, because the underlying left alignment code is not designed for or even remotely efficient at 1000+ base searches so it must be avoided somehow.

Reply:
At the very least I would like to see a test that demonstrates you can left align a 3 base str that extends over 50 bases on the reference properly with this approach, as it currently seems very few test cases actually hit this line.

Reply:
So I initially thought you were totally right about this.  But it turns out this auto-expansion actually does work, although it took me a little while to figure out why when the tests I wrote to demonstrate the problem unexpectedly passed.  Essentially, when an indel is not left aligned it can always be moved one base (and possibly more) to the left.  When we are testing whether an indel can be moved one base to the left in AlignmentUtils.leftAlignSingleIndel, we are testing whether the sequence can be represented by an indel of the same length placed one base to the left, not whether the sequence can be represented by the same inserted/deleted bases placed one base to the left.  So even when the str is not a factor of 50, the indel will still be initially aligned to the beginning of the given sequence, leading to the window being expanded.  I do agree that there are no tests showing that this works correctly though, so I will add some.